### PR TITLE
Implementation of session initialisation

### PR DESCRIPTION
**Pull request by lolczak**
### Pull Request Description
This PR enables a client to initialize the connection used to send the textual protocol messages.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/protocol/ClientController.scala [hunk 1]
```diff
@@ -18,19 +18,27 @@ import org.enso.languageserver.filemanager.FileManagerApi._
...
+import org.enso.languageserver.util.UnhandledLogging
 
 import scala.concurrent.duration._
 
 /**
   * An actor handling communications between a single client and the language
   * server.
   *
-  * @param clientId the internal client id.
+  * @param connectionId the internal client id.
```
Comment (4e6) [inline engine/language-server/src/main/scala/org/enso/languageserver/protocol/ClientController.scala:41]:
```suggestion
  * @param connectionId the internal connection id.
```

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala [hunk 2]
```diff
@@ -1733,7 +1742,7 @@ class TextOperationsTest extends BaseServerTest {
 
     "persist changes from a buffer to durable storage" in {
       val client = new WsTestClient(address)
-
+      initSession(client)
```
Comment (4e6) [inline engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala:1745]: `initSession` call is quite repetitive. I think it is worth creating two smart constructors for `WsTestClient`, returning initialized (as an `apply` method?) and uninitialized client.

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/ContextRegistryTest.scala [hunk 3]
```diff
@@ -11,7 +11,7 @@ class ContextRegistryTest extends BaseServerTest {
 
     "create execution context" in {
       val client = new WsTestClient(address)
-
+      initSession(client)
```
Comment (iamrecursion) [inline engine/language-server/src/test/scala/org/enso/languageserver/websocket/ContextRegistryTest.scala:14]: None of these tests are actually testing the initialisation. You should refactor this pattern into a testing helper function.

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/FileManagerTest.scala [hunk 4]
```diff
@@ -12,7 +12,7 @@ class FileManagerTest extends BaseServerTest {
 
     "write textual content to a file" in {
       val client = new WsTestClient(address)
-
+      initSession(client)
```
Comment (iamrecursion) [inline engine/language-server/src/test/scala/org/enso/languageserver/websocket/FileManagerTest.scala:15]: Same as above for here.

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/MonitoringTest.scala [hunk 5]
```diff
@@ -22,6 +22,7 @@ class MonitoringTest extends BaseServerTest {
           }
           """)
     }
+
```
Comment (iamrecursion) [inline engine/language-server/src/test/scala/org/enso/languageserver/websocket/MonitoringTest.scala:25]: These blank lines shouldn't be here.

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/ReceivesTreeUpdatesHandlerTest.scala [hunk 6]
```diff
@@ -10,14 +10,14 @@ class ReceivesTreeUpdatesHandlerTest extends BaseServerTest {
 
     "acquire capability receivesTreeUpdates" in {
       val client = new WsTestClient(address)
-
+      initSession(client)
```
Comment (iamrecursion) [inline engine/language-server/src/test/scala/org/enso/languageserver/websocket/ReceivesTreeUpdatesHandlerTest.scala:13]: This pattern should be refactored here too. Given it occurs across a lot of tests, I would suggest adding it as `getInitialisedWsClient` to a `LanguageServerTest` trait, much like the runtime has `CompilerTest` and `InterpreterTest` that contain common utility functions.

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala [hunk 7]
```diff
@@ -13,7 +13,7 @@ class TextOperationsTest extends BaseServerTest {
       // 1. Client tries to open a non-existent file.
       // 2. Client receives an error message.
       val client = new WsTestClient(address)
-
+      initSession(client)
```
Comment (iamrecursion) [inline engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala:16]: Here too.

**Comment by iamrecursion** [comment]
Also don't forget to remove the `session/end` message from the spec document.

File: doc/design/engine/engine-services.md [hunk 8]
```diff
@@ -2836,3 +2837,23 @@ Signals that requested capability is not acquired.
...
+
+##### `SessionNotInitialisedError`
+Signals that requested cannot be proccessed, beacuse session is not initialised.
+
+```typescript
+"error" : {
+  "code" : 6001,
+  "message" : "Session not initialised"
+}
+
```
[the remark this answers is not in the document]
Comment (lolczak) [reply doc/design/engine/engine-services.md]: fixed

File: doc/design/engine/engine-services.md [hunk 9]
```diff
@@ -2836,3 +2837,23 @@ Signals that requested capability is not acquired.
   "code" : 5001,
   "message" : "Capability not acquired"
 }
+
```
[the remark this answers is not in the document]
Comment (lolczak) [reply doc/design/engine/engine-services.md]: fixed

File: doc/design/engine/engine-services.md [hunk 10]
```diff
@@ -2836,3 +2837,23 @@ Signals that requested capability is not acquired.
...
+
+##### `SessionAlreadyInitialisedError`
+Signals that session is already initialised.
+
+```typescript
+"error" : {
+  "code" : 6002,
+  "message" : "Session already initialised"
+}
```
[the remark this answers is not in the document]
Comment (lolczak) [reply doc/design/engine/engine-services.md:2859]: fixed

File: engine/language-server/src/main/scala/org/enso/languageserver/session/SessionApi.scala [hunk 11]
```diff
@@ -0,0 +1,35 @@
+package org.enso.languageserver.session
+
+import java.util.UUID
+
+import org.enso.jsonrpc.{Error, HasParams, HasResult, Method}
+
+/**
+  * The connection management JSON RPC API provided by the language server.
+  * See [[https://github.com/luna/enso/blob/master/doc/design/engine/engine-services.md]]
```
[the remark this answers is not in the document]
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/session/SessionApi.scala]: fixed

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/ContextRegistryTest.scala (hunk 3 above)
Comment (lolczak) [reply engine/language-server/src/test/scala/org/enso/languageserver/websocket/ContextRegistryTest.scala]: fixed

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/FileManagerTest.scala (hunk 4 above)
Comment (lolczak) [reply engine/language-server/src/test/scala/org/enso/languageserver/websocket/FileManagerTest.scala]: fixed

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/MonitoringTest.scala (hunk 5 above)
Comment (lolczak) [reply engine/language-server/src/test/scala/org/enso/languageserver/websocket/MonitoringTest.scala]: fixed

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/ReceivesTreeUpdatesHandlerTest.scala (hunk 6 above)
Comment (lolczak) [reply engine/language-server/src/test/scala/org/enso/languageserver/websocket/ReceivesTreeUpdatesHandlerTest.scala]: fixed

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala (hunk 7 above)
Comment (lolczak) [reply engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala]: fixed

File: engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala (hunk 2 above)
Comment (lolczak) [reply engine/language-server/src/test/scala/org/enso/languageserver/websocket/TextOperationsTest.scala]: fixed

**Pull request merged**